fendermint/vm/interpreter/src/fvm/gas/actor.rs

```diff
@@ -1,123 +1,105 @@
+use fendermint_actor_gas_market::GasMarketReading;
+use fendermint_vm_actor_interface::gas::GAS_MARKET_ACTOR_ADDR;
 use fendermint_vm_actor_interface::system;
-use fvm_ipld_blockstore::Blockstore;
-use fvm_ipld_encoding::BytesDe;
-use std::sync::atomic::{AtomicU64, Ordering};
+use fvm::executor::{ApplyKind, Executor};
+use fvm_shared::clock::ChainEpoch;
 
-type AtomicGas = AtomicU64;
-type GasMarketState = fendermint_actor_gas_market::EIP1559GasState;
-
-/// The gas market based on EIP1155
-/// Due to the reference trait bound limit (`&self` instead of `&mut self`) in Interpreter, `Atomic`
-/// is used. However, the calling pattern should be single threaded, so direct `store` could be used.
-/// The usage of `Atomic` is purely to bypass the compilation issue without using unsafe.
-/// TODO: remove this overhead when trait bound is updated.
 #[derive(Default)]
 pub struct ActorGasMarket {
     /// The block gas limit
-    block_gas_limit: AtomicGas,
+    block_gas_limit: Gas,
     /// The accumulated gas usage so far
-    block_gas_used: AtomicGas,
+    block_gas_used: Gas,
 }
 
 impl GasMarket for ActorGasMarket {
     fn available_block_gas(&self) -> Gas {
-        self.block_gas_limit.load(Ordering::SeqCst) - self.block_gas_used.load(Ordering::SeqCst)
+        self.block_gas_limit - self.block_gas_used
     }
 
-    fn consume_gas(&self, gas: Gas) -> anyhow::Result<()> {
-        let block_gas_used = self.block_gas_used.load(Ordering::SeqCst);
-
-        if block_gas_used + gas >= self.block_gas_limit.load(Ordering::SeqCst) {
-            anyhow::bail!("out of block gas")
+    fn record_gas_used(&mut self, gas: Gas) -> anyhow::Result<()> {
+        if self.block_gas_used + gas >= self.block_gas_limit {
+            tracing::warn!("out of block gas, should not have happened")
         }
-        self.block_gas_used
-            .store(block_gas_used + gas, Ordering::SeqCst);
+        self.block_gas_used = self.block_gas_used.saturating_add(gas);
 
         Ok(())
     }
+}
 
-    fn update_params<DB: Blockstore + Clone + 'static>(
-        &self,
-        chain_state: &mut FvmExecState<DB>,
-    ) -> anyhow::Result<()> {
-        let block_gas_used = self.block_gas_used.load(Ordering::SeqCst);
-        let params = fvm_ipld_encoding::RawBytes::serialize(block_gas_used)?;
-
+impl ActorGasMarket {
+    pub fn new<E: Executor>(
+        executor: &mut E,
+        block_height: ChainEpoch,
+    ) -> anyhow::Result<ActorGasMarket> {
         let msg = FvmMessage {
             from: system::SYSTEM_ACTOR_ADDR,
-            to: GAS_ACTOR_ADDR,
-            sequence: chain_state.block_height() as u64,
+            to: GAS_MARKET_ACTOR_ADDR,
+            sequence: block_height as u64,
             // exclude this from gas restriction
-            gas_limit: u64::MAX,
-            method_num: fendermint_actor_gas_market::Method::UpdateBlockGasConsumption as u64,
-            params,
+            gas_limit: fvm_shared::BLOCK_GAS_LIMIT,
```

Comment:
I think other parts use max u64 for implicit executions?

Reply:
But it has to be `i64::MAX`, otherwise an error is thrown. But I think `BLOCK_GAS_LIMIT` should be fine for this case, should not be very gas costly.

Reply:
Using this value is incorrect. It is neither the gas limit configured for the subnet, nor MAX. It's just another arbitrary value inherited from Filecoin. I'll push a commit to switch to `i64::MAX`. Implicit executions are critical system executions; they cannot fail ever, so we must give them unbounded gas.

Reply:
For example, Lotus uses `block_gas_limit * 10000` for Cron execution: 

https://github.com/filecoin-project/lotus/blob/a5cb674bb11d3a438e5b9d9ca3da218f2941a7ac/chain/consensus/compute_state.go#L132